### PR TITLE
Fix S4507 FP: Top-level statements

### LINKED ISSUE #6772: Fix S4507 FP: Error raised on .NET 7 although the debug feature is deactivated
### Description

I migrate a project to .NET 7. The error S4507 is raised although the debug feature is desactivated in production mode.

### Repro steps

Code inside ``program.cs`` : 

```c#
var builder = WebApplication.CreateBuilder(args);
```

```c#
var app = builder.Build();

app.UseExceptionHandler("/error");
if (app.Environment.IsDevelopment())
    app.UseDeveloperExceptionPage();
else
    app.UseHsts();
```

Link to the SonarCloud public project where the issue appears : [link](https://sonarcloud.io/project/security_hotspots?id=VianneyDoleans_BlogCore)

### Expected behavior

Code is already deactivated so it should not be highlighted

### Actual behavior

 The error S4507 is raised.

### Related information

* .NET 7
* I suppose the problem is related to [minimal hosting model introduced with .NET 6](https://learn.microsoft.com/en-us/aspnet/core/migration/50-to-60-samples?view=aspnetcore-7.0).

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #6772

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs [hunk 1]
```diff
@@ -30,9 +30,15 @@ public sealed class DeliveringDebugFeaturesInProduction : DeliveringDebugFeature
         internal /*for testing*/ DeliveringDebugFeaturesInProduction(IAnalyzerConfiguration configuration) : base(configuration) { }
 
         // For simplicity we will avoid creating noise if the validation is invoked in the same method (https://github.com/SonarSource/sonar-dotnet/issues/5032)
-        protected override bool IsDevelopmentCheckInvoked(SyntaxNode node, SemanticModel semanticModel) =>
-            node.FirstAncestorOrSelf<MethodDeclarationSyntax>() is { } parentMethodDeclaration
-            && parentMethodDeclaration.DescendantNodes().Any(x => IsDevelopmentCheck(x, semanticModel));
+        protected override bool IsDevelopmentCheckInvoked(SyntaxNode node, SemanticModel semanticModel)
+        {
+            SyntaxNode parentMethod = node.FirstAncestorOrSelf<MethodDeclarationSyntax>();
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:35]:
```suggestion
            SyntaxNode parentMethod = node.FirstAncestorOrSelf<BaseMethodDeclarationSyntax>();
```

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:35]: Misses local functions, lambdas, property getter/setter, indexer getter/setter, event adder/remover, field initializer. This should look more like this:
https://github.com/SonarSource/sonar-dotnet/blob/71a67baddd08c31c8c97365b7bb74c66b397e1b5/analyzers/src/SonarAnalyzer.CSharp/Rules/NonAsyncTaskShouldNotReturnNull.cs#L63-L77

and should be an extension method: `GetContainer` or `GetEnclosingScope`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs [hunk 2]
```diff
@@ -30,9 +30,15 @@ public sealed class DeliveringDebugFeaturesInProduction : DeliveringDebugFeature
...
+            if (parentMethod is null && node.FirstAncestorOrSelf<GlobalStatementSyntax>() is { }) // for top-level statements
+            {
+                parentMethod = node.FirstAncestorOrSelf<CompilationUnitSyntax>();
+            }
+            return parentMethod is not null && parentMethod.DescendantNodes().Any(x => IsDevelopmentCheck(x, semanticModel));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:40]: This result in a lot of FNs: 
```cs
{
    app.Environment.IsDevelopment();
    app.UseDeveloperExceptionPage();
}
```
Do we have tests for that?

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs (hunk 2 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs]: [Yes, it's covered.](https://github.com/SonarSource/sonar-dotnet/blob/cff1e6ade43eece40049467405aab5b3307cf91a/analyzers/tests/SonarAnalyzer.Test/TestCases/Hotspots/DeliveringDebugFeaturesInProduction.NetCore2.cs#L123)
There are several FNs. The rule is pretty lenient about this. It doesn't involve symbolic execution or use the control flow in any way. It just checks whether there's a call to the `IsDevelopment()` method in the same scope (regardless of where that call is located).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs [hunk 3]
```diff
@@ -42,5 +42,20 @@ public sealed class DeliveringDebugFeaturesInProduction : DeliveringDebugFeature
...
+                SyntaxKind.DestructorDeclaration,
+                SyntaxKind.GetAccessorDeclaration,
+                SyntaxKind.GlobalStatement,
+                SyntaxKindEx.InitAccessorDeclaration,
+                SyntaxKindEx.LocalFunctionStatement,
+                SyntaxKind.MethodDeclaration,
+                SyntaxKind.ParenthesizedLambdaExpression,
+                SyntaxKind.RemoveAccessorDeclaration,
+                SyntaxKind.SetAccessorDeclaration,
+                SyntaxKind.SimpleLambdaExpression));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:59]: `IsAnyKind` allocates an array on each invocation and should not be used anymore (see e.g. https://github.com/SonarSource/sonar-dotnet/issues/8103)
```suggestion
        private static SyntaxNode EnclosingScope(SyntaxNode node) =>
            node.Ancestors().FirstOrDefault(x => x.Kind() is SyntaxKind.AddAccessorDeclaration
                or SyntaxKind.ConstructorDeclaration
                or SyntaxKind.DestructorDeclaration
                or SyntaxKind.GetAccessorDeclaration
                or SyntaxKind.GlobalStatement
                or SyntaxKindEx.InitAccessorDeclaration
                or SyntaxKindEx.LocalFunctionStatement
                or SyntaxKind.MethodDeclaration
                or SyntaxKind.ParenthesizedLambdaExpression
                or SyntaxKind.RemoveAccessorDeclaration
                or SyntaxKind.SetAccessorDeclaration
                or SyntaxKind.SimpleLambdaExpression));
```

Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:59]: Initializer blocks are missing (SyntaxKind.EqualsValueClause)
```cs
public class C
{
    int i = 0;
    int I { get; } = 0;
}
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs [hunk 4]
```diff
@@ -42,5 +42,20 @@ public sealed class DeliveringDebugFeaturesInProduction : DeliveringDebugFeature
         private bool IsDevelopmentCheck(SyntaxNode node, SemanticModel semanticModel) =>
             node is InvocationExpressionSyntax condition
             && IsValidationMethod(semanticModel, condition, condition.Expression.GetIdentifier()?.ValueText);
+
+        private static SyntaxNode EnclosingScope(SyntaxNode node) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:46]: Move the method to SonarAnalyzer.CSharp -> SyntaxNodeExtensions (and ideally bring it under test for all kinds of code blocks).

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs (hunk 3 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs:59]: And three more occasions where we have expressions that do not belong to a method directly:
* [ConstructorInitializerSyntax](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.constructorinitializersyntax?view=roslyn-dotnet-4.7.0)
* [PrimaryConstructorBaseTypeSyntax](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.primaryconstructorbasetypesyntax?view=roslyn-dotnet-4.7.0)
 
[sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yANDCFMjAD4ACAzAAScAmXgBEAsACgA3hN6z+fYQAoipGLyIBKXpN4BfCfvESe/IehAiJ08XPm90ijRZgALIskUBGXgGpenrQB6QP9fUPd1UnUYAHJkXkwAdyj8MmoADxgGXmo2XjBMBF4AQQQAczgAWwhVAGUAT1UodIA6AFF0gAckZGQSKMSqF14ABSgkVRaxibUIrFI2BDhCQoBJUioiOiIALxoGpvTeGTkdQ1sTBxVZp15gVghlHx8gwKJn9XiVSJh4pJS0hBMidZGcDBJjHwkKkEAATXgAJWUql4qw0AG4IeITNDCvDasi1GiLEjVs8MbJgmS/BEyNE4glkrxUqogTAgA)

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 5]
```diff
@@ -19,27 +19,45 @@
...
 namespace SonarAnalyzer.Extensions
 {
     internal static partial class SyntaxNodeExtensions
     {
         private static readonly ControlFlowGraphCache CfgCache = new();
-        private static readonly SyntaxKind[] ParenthesizedNodeKinds = new[] { SyntaxKind.ParenthesizedExpression, SyntaxKindEx.ParenthesizedPattern };
+        private static readonly SyntaxKind[] ParenthesizedNodeKinds = [SyntaxKind.ParenthesizedExpression, SyntaxKindEx.ParenthesizedPattern];
+        private static readonly SyntaxKind[] EnclosingScopeSyntaxKinds = [
+            SyntaxKind.AddAccessorDeclaration,
+            SyntaxKind.AddAccessorDeclaration,
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:31]:
```suggestion
            SyntaxKind.AddAccessorDeclaration,
```

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 6]
```diff
@@ -61,12 +79,12 @@ public static bool IsPartOfBinaryNegationOrCondition(this SyntaxNode node)
...
-                                              SyntaxKind.WhileStatement,
-                                              SyntaxKind.ConditionalExpression,
-                                              SyntaxKind.MethodDeclaration,
-                                              SyntaxKind.SimpleLambdaExpression) ?? false)
+            while (current.Parent != null && current.Parent?.Kind() is not (SyntaxKind.BitwiseNotExpression
+                                                                     or SyntaxKind.IfStatement
+                                                                     or SyntaxKind.WhileStatement
+                                                                     or SyntaxKind.ConditionalExpression
+                                                                     or SyntaxKind.MethodDeclaration
+                                                                     or SyntaxKind.SimpleLambdaExpression))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:87]: Misses
or SyntaxKind.AnonymousMethodExpression
or SyntaxKind.ParenthesizedLambdaExpression

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 7]
```diff
@@ -19,27 +19,45 @@
...
+            SyntaxKind.GlobalStatement,
+            SyntaxKindEx.InitAccessorDeclaration,
+            SyntaxKindEx.LocalFunctionStatement,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.ParenthesizedLambdaExpression,
+            SyntaxKindEx.PrimaryConstructorBaseType,
+            SyntaxKind.RemoveAccessorDeclaration,
+            SyntaxKind.SetAccessorDeclaration,
+            SyntaxKind.SimpleLambdaExpression,
+            SyntaxKind.ThisConstructorInitializer];
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:47]: Misses
OperatorDeclaration, ConversionOperatorDeclaration

File: analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs [hunk 8]
```diff
@@ -1200,6 +1200,49 @@ public class Derived(int i) {{baseType}} { }
...
+        [DataRow("""int Property { set { $$_ = value;$$ } }""", SyntaxKind.SetAccessorDeclaration)]
+        [DataRow("""int Property { set { $$_ = value;$$ } }""", SyntaxKind.SetAccessorDeclaration)]
+        [DataRow("""int Method() { return LocalFunction(); int LocalFunction() { $$return 42;$$ } }""", SyntaxKindEx.LocalFunctionStatement)]
+        [DataRow("""int Method() { return LocalFunction(); int LocalFunction() => $$42$$; }""", SyntaxKindEx.LocalFunctionStatement)]
+        [DataRow("""int Method() { $$return 42;$$ }""", SyntaxKind.MethodDeclaration)]
+        [DataRow("""int Method() => $$42$$;""", SyntaxKind.MethodDeclaration)]
+        [DataRow("""int Method() { var lambda = () => $$42$$; return lambda(); }""", SyntaxKind.ParenthesizedLambdaExpression)]
+        [DataRow("""int Method() { Func<int, int> lambda = x => $$x + 1$$; return lambda(42); }""", SyntaxKind.SimpleLambdaExpression)]
+        [DataRow("""event EventHandler SomeEvent { add { int x = 42; } remove { $$int x = 42;$$ } }""", SyntaxKind.RemoveAccessorDeclaration)]
+        [DataRow("""Derived(int arg) : this($$arg.ToString()$$) { }""", SyntaxKind.ThisConstructorInitializer)]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Extensions/SyntaxNodeExtensionsTest.cs:1220]: Misses tests for 
OperatorDeclaration, ConversionOperatorDeclaration

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good. Two more code block syntax kinds are missing

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**